enclave-manager/server/server.go

```diff
@@ -241,14 +237,155 @@ func (c *WebServer) GetServices(ctx context.Context, req *connect.Request[kurtos
 		return nil, err
 	}
 
+	serviceInfoMapFromApicObj := serviceInfoMapFromAPIC.Msg.GetServiceInfo()
+
+	// we aren't in a cloud context, so we exit early
+	if !c.enforceAuth {
+		resp := &connect.Response[kurtosis_core_rpc_api_bindings.GetServicesResponse]{
+			Msg: &kurtosis_core_rpc_api_bindings.GetServicesResponse{
+				ServiceInfo: serviceInfoMapFromApicObj,
+			},
+		}
+		return resp, nil
+	}
+
+	cloudClient, err := c.createKurtosisCloudBackendClient(
+		kurtosisCloudApiHost,
+		kurtosisCloudApiPort,
+	)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Failed to create the Cloud backend client")
+	}
+
+	jwtToken, err := extractJwtToken(req.Header())
+	if err != nil {
+		return nil, err
+	}
+	getUnlockedPortsRequest := &connect.Request[kurtosis_backend_server_rpc_api_bindings.GetUnlockedPortsRequest]{
+		Msg: &kurtosis_backend_server_rpc_api_bindings.GetUnlockedPortsRequest{
+			AccessToken:       jwtToken,
+			InstanceShortUuid: instanceConfig.InstanceId[:shortUuidLength],
+			EnclaveShortUuid:  req.Msg.EnclaveShortenedUuid,
+		},
+	}
+
+	var unlockedPortsAndServices []*kurtosis_backend_server_rpc_api_bindings.Port
+	getUnauthenticatedPortsResponse, err := (*cloudClient).GetUnlockedPorts(ctx, getUnlockedPortsRequest)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "an error occurred while pulling unauthenticated ports from the cloud backend")
+	}
+	unlockedPortsAndServices = getUnauthenticatedPortsResponse.Msg.Port
+
+	for _, service := range serviceInfoMapFromApicObj {
+		serviceShortUuid := service.ServiceUuid[:shortUuidLength]
+		for _, privatePort := range service.PrivatePorts {
+			locked := true
+			for _, unlockedPortsAndService := range unlockedPortsAndServices {
+				if unlockedPortsAndService.ServiceShortUuid == serviceShortUuid {
+					if privatePort.Number == unlockedPortsAndService.PortNumber {
+						locked = false
+						break
+					}
+				}
+			}
+			privatePort.Locked = &locked
+		}
+	}
```

Comment:
I am assuming this performs fine because users have a small number of services and ports.

Reply:
yes! not happy about this but our numbers are tiny < 100 mostly